File: src/lib/api/approval.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  ApprovalRequest,
  CreateApprovalRequestInput,
  DecideApprovalRequestInput,
  ListApprovalRequestsInput,
} from "@/types/approval";

export const approvalApi = {
  list(input: ListApprovalRequestsInput = {}): Promise<ApprovalRequest[]> {
    if (hasTauriRuntime()) {
      return invoke("list_approval_requests", { input });
    }
    const params = new URLSearchParams();
    if (input.status) params.set("status", input.status);
    if (input.limit) params.set("limit", String(input.limit));
    const query = params.toString();
    return devApiFetch(`/approvals${query ? `?${query}` : ""}`);
  },

  create(input: CreateApprovalRequestInput): Promise<ApprovalRequest> {
    if (hasTauriRuntime()) {
      return invoke("create_approval_request", { input });
    }
    return devApiFetch("/approvals", {
      method: "POST",
      body: JSON.stringify(input),
    });
  },

  decide(input: DecideApprovalRequestInput): Promise<ApprovalRequest> {
    if (hasTauriRuntime()) {
      return invoke("decide_approval_request", { input });
    }
    return devApiFetch("/approvals/decide", {
      method: "POST",
      body: JSON.stringify(input),
    });
  },
};

File: src-tauri/src/commands/approval.rs
use crate::error::CommandError;
use crate::models::{
    ApprovalRequest, CreateApprovalRequestInput, DecideApprovalRequestInput,
    ListApprovalRequestsInput,
};
use crate::services::approval::ApprovalService;
use crate::state::AppState;

#[tauri::command]
pub fn list_approval_requests(
    state: tauri::State<'_, AppState>,
    input: ListApprovalRequestsInput,
) -> Result<Vec<ApprovalRequest>, CommandError> {
    ApprovalService::list(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn create_approval_request(
    state: tauri::State<'_, AppState>,
    input: CreateApprovalRequestInput,
) -> Result<ApprovalRequest, CommandError> {
    ApprovalService::create(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn decide_approval_request(
    state: tauri::State<'_, AppState>,
    input: DecideApprovalRequestInput,
) -> Result<ApprovalRequest, CommandError> {
    ApprovalService::decide(&state.db, input).map_err(|e| e.into())
}
